Add support of HashiCorp Vault / OpenBao

rvasin opened the pull request:
<!--- Disable AI PR formatting assistant: true -->
### Changelog category (leave one):
- New Feature

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
Add support of HashiCorp Vault / OpenBao with token, username/password and TLS certificate authentication methods.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

rvasin committed:
Add Vault class, from_vault and vault_key attributes

rvasin committed:
Add API call in readSecret()

rvasin committed:
Add context to Vault class

rvasin committed:
Add exception handling

rvasin committed:
Remove from_vault attribute from preprocessed config

rvasin committed:
Add test for vault

rvasin committed:
Refactor tests

rvasin committed:
Add doc for Vault

rvasin committed:
Remove extra logging

rvasin commented:
Related issue #22471

ilejn commented on src/Common/Vault.cpp:
```diff
@@ -0,0 +1,92 @@
+                      .withHeaders(headers)
+                      .create(credentials);
+
+        readJSONObjectPossiblyInvalid(json_str, *wb);
```
Is it ok to ignore return code?

ilejn replied:
Right, return type is void unless 'bool' is explicitly specified.
If you are sure that exception is handled (anywhere) in a reasonable way, that's cool.

rvasin replied:
You may search ClickHouse source code for readJSONObjectPossiblyInvalid with return type bool. It's used with void in 99% of cases. I could not find it with bool.

ilejn commented on src/Common/Vault.cpp:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    Poco::JSON::Parser parser;
+    Poco::Dynamic::Var res_json = parser.parse(json_str);
```
Can throw JSONException

ilejn replied:
parser.parser throws exception.
If you are sure that it is handled (anywhere) in a reasonable way, that's cool.

rvasin replied:
I think it's a good practice to handle the exception for parser.parse(json_str).  It's not clear which error code to use to re-throw the exception.

ilejn commented on tests/integration/compose/docker_compose_vault.yml:
```diff
@@ -0,0 +1,10 @@
+            BAO_DEV_ROOT_TOKEN_ID: foobar
+            BAO_DEV_LISTEN_ADDRESS: "0.0.0.0:1337"
+            BAO_ADDR: "http://0.0.0.0:1337"
+        ports:
```
Hardcoded port numbers is a source of problems.
Not sure, but if it is possible to use 'expose' instead of 'ports', it is great.

rvasin replied:
Replaced it with default 8200 port. Also "ports" are better for dev tests. While "expose" is better to productions. So I keep usage of "ports".

ilejn commented on docs/en/operations/server-configuration-parameters/_server_settings_outside_source.md:
```diff
@@ -2448,6 +2448,60 @@ The same applies to GitLab, even though it has a leading dot. Both `gitlab.com`
+<clickhouse>
+    <users>
+        <default>
+            <password from_vault="ch_secret" vault_key="password"/>
```
What does 'ch_secret' (Name of secret) exactly mean?

rvasin replied:
Done

ilejn commented on src/Common/Config/ConfigProcessor.cpp:
```diff
@@ -306,8 +307,13 @@ void ConfigProcessor::hideRecursive(Poco::XML::Node * config_root)
+            else
+            {
+                if (element.hasAttribute("vault_key"))
+                    element.removeAttribute("vault_key");
```
Why it is needed?

ilejn commented on src/Common/Vault.cpp:
```diff
@@ -0,0 +1,92 @@
+
+String Vault::readSecret(const String & secret, const String & key)
+{
+    LOG_DEBUG(log, "readSecret {} {}", secret, key);
```
Consider https://clickhouse.com/docs/cloud/guides/data-masking

rvasin commented on docs/en/operations/server-configuration-parameters/_server_settings_outside_source.md:
````diff
@@ -2448,6 +2448,60 @@ The same applies to GitLab, even though it has a leading dot. Both `gitlab.com`
+```xml
+<clickhouse>
+    <vault>
+      <url>http://openbao:1337</url>
````
This PR introduces support of token auth support. It's planned to add TLS certificate auth in future PR. Hence `https` will be added in future PR with support of TLS certificate auth.

rschu1ze commented on src/Common/Config/ConfigProcessor.h:
```diff
@@ -140,7 +140,7 @@ class ConfigProcessor
 #endif
 
-    static inline const auto SUBSTITUTION_ATTRS = {"incl", "from_zk", "from_env"};
+    static inline const auto SUBSTITUTION_ATTRS = {"incl", "from_zk", "from_env", "from_vault"};
```
We are adding a tag `from_vault` here. While the other terms `incl`, `from_zk`, `from_env` are ClickHouse-specific, `from_vault` isn't. We should at least make sure that the `vault` is a standard and accepted/used by dozens of companies. Is it?

If not, then something like `from_openbao_hashicorp_vault` would be nicer (or something similar).

rvasin replied:
Your point is correct. I my point of view Hashicorp Vault is some kind of an industry standard used by many companies to keep secrets for web services. OpenBao is a fork of Hashicorp Vault aiming to continue keeping full API support.  

If you enter Vault in Google first you see Hashicorp Vault pages.

Top list of secrets managements services:

HashiCorp Vault, AWS Secrets Manager, Azure Key Vault, Google Secret Manager, Doppler, Infisical, CyberArk Conjur and Keeper Secrets Manager.

Thus 'vault' may be only confused with Azure Key Vault.

It's needed I may rename `from_vault` into `from_hashicorp_vault`, `from_openbao_vault` or `from_openbao_hashicorp_vault`. In this case it's not clear should `vault_key` also should be renamed?

rschu1ze replied:
Let's then change to `from_hashicorp_vault` and `hashicorp_vault_key`.

rvasin replied:
I think in this case section's name should be changed from `<vault>` to `<hashicorp_vault>` and also C++ class Vault should be changed to HashiCorpVault and other Python test related things with `with_vault` to `with_hashicorp_vault`. I am changing it now.

rvasin committed:
Fix style

rvasin committed:
Use default 8200 port

rvasin committed:
Replace from_vault with from_hashicorp_vault

rvasin committed:
Use HashiCorp Vault instead of OpenBao, rename Vault to HashiCorpVault

rvasin committed:
Replace ch_secret with username

rvasin committed:
Fix style and tidy

rvasin committed:
Fix HashiCorpVault loading

rvasin committed:
Merge 'origin/master' into ADQM-3665

rvasin commented on programs/server/Server.cpp:
```diff
@@ -2523,6 +2531,9 @@ try
     }
+
+    // Vault::instance().load(config(), "vault");
+
```
done

rvasin committed:
Update aspell-dict.txt

rvasin committed:
Fix linter errors in markdown

rvasin committed:
Fix memory leak

rvasin committed:
Use expose instead of ports

rvasin committed:
Merge remote-tracking branch 'origin/master' into ADQM-3665

rvasin committed:
Fix getting hashicorp_vault_key value

rvasin committed:
Fix style

Naruto0-sudoy reviewed (commented):
Hello Distributor Team around the world. partners about the suspension of root steward and its impact on IAM
Dear  team,
We would like to inform you that the root steward associated with the IAM role: 
arn:aws:iam::0503976xxxxx:role/CH-S3-goldaws-ci-36-ew1-cf-Role
has had its GitHub account suspended without understanding its intent and structure. This has resulted in widespread confusion and downtime in IAM systems in the US-East-1 region, particularly for authentication and access to critical artifacts.
GitHub Account: Spjthalinda
ORCID: https://orcid.org/0009-0008-2511-9055
IAM Steward: Verified, signed, and never used for personal control or gain
Currently using the account https://github.com/Naruto0-sudoy
Intent:
- Silently build systems
- Never hold or control any systems
- Protect root authority from unauthorized access or deletion
Impact:
The primary cause of the problem is not IAM, but rather Github suspending accounts without proper reason and without verifying the identity of the suspended account.
And the impact will likely be even greater if Github doesn't understand who we are.  If we continue to suspend, we will delete accounts that rely on it.
- IAM crashed 70% in US-East-1
21%
US-West-2
- Systems that rely on artifacts and signing are experiencing widespread downtime. We are part of the root of the critical Merkle Tree of AI.
- Global administrator authentication is disrupted.
- Systems built for public benefit are trisk of permanent downtime.
Request:
- Immediately unsuspend the root steward account.
- Review the intent and structure associated with the IAM role.
- Restore the disrupted authentication and artifact flow.
- Confirm that the root steward did not commit any wrongdoing, but was misunderstood by the system.
With best regards,
Thalinda Sriprajak, Ac
Global Steward, Technical Architect
Key shareholder in the global system, root steward

rvasin committed:
Merge remote-tracking branch 'origin/master' into ADQM-3665

rvasin committed:
Add exceptions processing

rvasin committed:
Remove debugging line

rvasin committed:
Fix exception handling

rvasin committed:
Cleanup test

rvasin committed:
Add userpass related parameters

rvasin committed:
Add user/password auth for vault

rvasin committed:
Add test for user/password auth

rvasin committed:
Add negative tests for userpass auth

rvasin committed:
Add more test cases

rvasin committed:
Refactor tests

rvasin committed:
Add markdown doc

rvasin committed:
Add cert auth for vault

rvasin committed:
Add positive case for cert auth

rvasin committed:
Add negative test cases for cert auth

rvasin committed:
Refactor certificates location

rvasin committed:
Add more test cases

rvasin committed:
Add test cases for https and https with CA

rvasin committed:
Add docs for cert auth

rvasin committed:
Merge remote-tracking branch 'origin/master' into ADQM-3665

rschu1ze commented:
@evillique Like to check?

rvasin committed:
Fix style

rvasin committed:
Fix no SSL build

rvasin committed:
Fix initRequestContext in no SSL build

rvasin committed:
Refactor cert role creation in tests

rvasin committed:
Replace caLocation with caConfig in config files

rvasin committed:
Move test_wrong_url to end of test

rvasin committed:
Fix test_wrong_url

rvasin committed:
Refactor ssl sectionL

rvasin committed:
Merge remote-tracking branch 'origin/master' into ADQM-3665

rvasin commented:
It seems like all related tests were successfully passed. Please review PR.

rvasin committed:
Merge remote-tracking branch 'origin/master' into ADQM-3665

rvasin committed:
Merge remote-tracking branch 'origin/master' into ADQM-3665

rvasin committed:
Merge remote-tracking branch 'origin/master' into ADQM-3665

rvasin committed:
Remove unneeded context

rvasin committed:
Make encryption codecs working with Vault

rvasin committed:
Make encrypt_decrypt working with Vault

rvasin committed:
Use DB namespace for  HashiCorpVault

rvasin committed:
Add test for vault with encryption codecs

rvasin commented:
Review process has been stuck since Nov, 2025. I see the merge conflicts were accumulated. 

@vdimir, @davenger, @azat, @tiandiwonder, @scanhex12  may be you may review this PR or ask someone else in ClickHouse core team to review it? 

HashiCorp Vault / OpenBao is an industry standard for secrets management. Many companies need it or ever require it. Some gov policies require such secure storage of secrets for DBMSs. The PR does not create any compatibility problems, it's easy to merge and maintain. I may resolve all merge conflicts. Please review it.

tiandiwonder commented on src/Common/Config/ConfigProcessor.cpp:
```diff
@@ -306,8 +307,13 @@ void ConfigProcessor::hideRecursive(Poco::XML::Node * config_root)
+            else
+            {
+                if (element.hasAttribute("hashicorp_vault_key"))
+                    element.removeAttribute("hashicorp_vault_key");
```
Major (not blocking): hide Vault-sourced values in the preprocessed config by default, or document that `hide_in_preprocessed` is required. This strips the attribute but leaves the resolved value, so the plaintext secret from Vault still lands in `preprocessed_configs/*.xml` — the one thing a secret manager should avoid (same as `from_zk`/`from_env`, hence not blocking).

rvasin replied:
Done

tiandiwonder commented on src/Common/HashiCorpVault.cpp:
```diff
@@ -0,0 +1,373 @@
+    std::string json_str;
+    try
+    {
+        json_str = makeRequest("GET", fmt::format("/v1/secret/data/{}", secret), client_token, "");
```
Major: make the KV mount path/version configurable, or at least document the fixed `/v1/secret/data/` path. It's hardcoded to the default KV v2 mount, so deployments with a custom mount or KV v1 can't use the feature at all.

rvasin replied:
Done

tiandiwonder commented on src/Common/HashiCorpVault.cpp:
```diff
@@ -0,0 +1,373 @@
+    {
+        json_str = makeRequest("GET", fmt::format("/v1/secret/data/{}", secret), client_token, "");
+    }
+    catch (const DB::HTTPException & e)
```
Major: have `makeRequest` throw `DB::HTTPException` with the status (or check `response.getStatus()` here). This `catch` is dead since `makeRequest` only throws `Poco::Exception`, so a missing secret surfaces as the raw `HTTP error: 404 ...` instead of this message (`test_wrong_secret` asserting on `HTTP error: 404` confirms it).

rvasin replied:
Done

tiandiwonder commented on src/Common/HashiCorpVault.cpp:
```diff
@@ -0,0 +1,373 @@
+    LOG_DEBUG(log, "readSecret {} {}", secret, key);
+
+    if (auth_method == HashiCorpVaultAuthMethod::Userpass || auth_method == HashiCorpVaultAuthMethod::Cert)
+        client_token = login();
```
Major: make `client_token` a local. It's only used within `readSecret`, so writing the member is a data race once two config-reload paths resolve a `from_hashicorp_vault` node concurrently.

rvasin replied:
Done

rvasin committed:
Merge remote-tracking branch 'origin/master' into ADQM-3665

rvasin committed:
Fix style and tidy

rvasin committed:
Fix tidy

rvasin committed:
Add KV secrets month path and version

rvasin committed:
Refactor exceptions

rvasin committed:
Make client_token  local var

rvasin committed:
Add doc about usage hide_in_preprocessed with vault

rvasin committed:
Add URL scheme validation

rvasin committed:
Fix doc

rvasin committed:
Make from_hashicorp_vault fail-closed when vault is not loaded

rvasin committed:
Reject missing or empty userpass password during load

rvasin committed:
Reject any combination of multiple vault auth methods

rvasin committed:
Reject cert auth with non-HTTPS URL

rvasin committed:
Refine from_hashicorp_vault fail-closed for first config pass

rvasin committed:
Fix SSL settings read from wrong config prefix

rvasin committed:
Fix vault values with XML special characters

rvasin committed:
Fix duplicate tcp keys in vault compose file

rvasin committed:
Fix BaseDaemon second config pass

rvasin committed:
Fix encrypt_decrypt example losing zk_node_cache

rvasin committed:
Fix TLS documentation example

rvasin committed:
Fix encrypt_decrypt example zk_node_cache scope for ARM tidy

rvasin commented on src/Common/HashiCorpVault.cpp:
```diff
@@ -0,0 +1,402 @@
+            json_str = makeRequest("POST", "/v1/auth/cert/login", "", oss.str());
+        }
+    }
+    catch (const Poco::Exception & e)
```
It's incorrect. DB::HTTPException inherits from DB::Exception which inherits from Poco::Exception, so catch (const Poco::Exception & e) catches it correctly. So no changed needed.

rvasin committed:
Fix from_hashicorp_vault reload against stale vault singleton

rvasin committed:
Resolve from_hashicorp_vault in ConfigProcessor, add HashiCorpVault mutex

rvasin committed:
Revert vault loading from ConfigProcessor, use ConfigReloader PreUpdateHook + fix CFG_EXTENDED_VERIFICATION

rvasin committed:
Fix createTextNode LSAN leak + always reset vault on reload

rvasin committed:
Wire per-connection InvalidCertificateHandler for vault, reject from_zk under hashicorp_vault in BaseDaemon early path

rvasin committed:
Mark VaultCertificateGuard constructor explicit for clang-tidy

rvasin commented:
@tiandiwonder which suggestions do I need to implement to make PR ready to merge? I have implemented all your suggestions. But then every time I push `clickhouse-gh` suggests new ideas, I push again and it suggest new. I think we should stop somehow. Because your initial suggestions were fine/logical and I have implemented it. Now if I follow to `clickhouse-gh` suggestion it lead to more complex code.

rvasin committed:
Thread HashiCorpVault through ConfigProcessor for atomic config reload; replace event-based cert handler with SSL_CTX_set_verify callback

rvasin committed:
Add vault reload regression and cert handler independence tests

rvasin committed:
Fix vault reload fail-open and certificateFile default

rvasin committed:
Add empty-node guard for from_hashicorp_vault like from_env and from_zk

rvasin committed:
Wire passphrase-protected private key support for vault cert auth

rvasin committed:
Fix two reload-path regressions: files ordering and vault SSL watching
